### PR TITLE
Pal 727 kafka zookeeper volume permissions

### PR DESCRIPTION
**Pull request by pd104923**
The changes for ResourceRequest and ResourceResponse were done as a SonarQube code smell came about after a develop merge.

### CODE REVIEW

File: charts/kafka/values.yaml [hunk 1]
```diff
@@ -42,8 +42,11 @@ resources: {}
   #   memory: 1024Mi
 kafkaHeapOptions: "-Xmx1G -Xms1G"
 
-## Optional Container Security context
-securityContext: {}
+## Kafka pod Security Context
+securityContext:
+  enabled: true
+  fsGroup: 0
+  runAsUser: 0
```
Comment (dev930018) [inline charts/kafka/values.yaml:49]: I'm happy with this, the confluentinc kafka image requires root, and we can't change that.
If this had to be changed (to be in-line with redis, zookeeper and using runAsUser 1001 etc...), we should switch to the bitnami image.

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceRequest.java [hunk 2]
```diff
@@ -44,13 +44,16 @@
...
     // Json Node representation of Context
     private final JsonNode context;
     // Json Node representation of the User
     private final JsonNode user;
 
+    /**
+     * Resource Id of the resource requested
+     */
+    public final String resourceId;
+
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceRequest.java:56]: SonarQube?

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceResponse.java [hunk 3]
```diff
@@ -52,7 +52,10 @@
     private final JsonNode context;
     // Json Node representation of the User
     private final JsonNode user;
-    // Resource that has been requested to access
+
+    /**
+     * Resource that has been requested to access
+     */
     public final LeafResource resource;
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceResponse.java:59]: More SonarQube?

**Review by dev930018 — DISMISSED** [review]
Whether or not the SQ code-smells should be fixed here is up for debate.

**Review by kg981167224 — DISMISSED** [review]

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceRequest.java (hunk 2 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceRequest.java:56]: Yep. Became a smell after a develop merge, somehow. Thought I'd just fix it.

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceResponse.java (hunk 3 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/request/ResourceResponse.java:59]: Yep. Same reasoning.

File: charts/kafka/values.yaml (hunk 1 above)
Comment (pd104923) [reply charts/kafka/values.yaml:49]: Might be a consideration for the future if we don't want to run as root.

**Review by m78233 — DISMISSED** [review]

File: charts/kafka/charts/zookeeper/templates/_helpers.tpl [hunk 4]
```diff
@@ -40,6 +40,29 @@ If release name contains chart name it will be used as a full name.
...
 
+{{/*
+Return the proper image name (for the init container volume-permissions image)
+*/}}
+{{- define "zookeeper.volumePermissions.image" -}}
+{{- $registryName := .Values.volumePermissions.image.registry -}}
+{{- $repositoryName := .Values.volumePermissions.image.repository -}}
+{{- $tag := .Values.volumePermissions.image.tag | toString -}}
+{{/*
+Helm 2.11 supports the assignment of a value to a variable defined in a different scope,
```
Comment (dev-958) [inline charts/kafka/charts/zookeeper/templates/_helpers.tpl]: Can we get shot of this and support only helm 3+ ?

File: charts/kafka/charts/zookeeper/values.yaml [hunk 5]
```diff
@@ -244,6 +246,27 @@ exporters:
...
+    tag: buster
+    pullPolicy: IfNotPresent
+    ## Optionally specify an array of imagePullSecrets.
+    ## Secrets must be manually created in the namespace.
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+    ##
+    # pullSecrets:
+    #   - myRegistryKeySecretName
+  resources: {}
+  # resources:
```
Comment (dev-958) [inline charts/kafka/charts/zookeeper/values.yaml]: might want to clean-up these comments as they are not useful?

File: charts/kafka/templates/_helpers.tpl [hunk 6]
```diff
@@ -53,6 +53,29 @@ We truncate at 63 chars because some Kubernetes name fields are limited to this
...
 
+{{/*
+Return the proper image name (for the init container volume-permissions image)
+*/}}
+{{- define "kafka.volumePermissions.image" -}}
+{{- $registryName := .Values.volumePermissions.image.registry -}}
+{{- $repositoryName := .Values.volumePermissions.image.repository -}}
+{{- $tag := .Values.volumePermissions.image.tag | toString -}}
+{{/*
+Helm 2.11 supports the assignment of a value to a variable defined in a different scope,
```
Comment (dev-958) [inline charts/kafka/templates/_helpers.tpl]: ditto from above

File: charts/kafka/values.yaml [hunk 7]
```diff
@@ -288,6 +291,27 @@ jmx:
...
+    tag: buster
+    pullPolicy: IfNotPresent
+    ## Optionally specify an array of imagePullSecrets.
+    ## Secrets must be manually created in the namespace.
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+    ##
+    # pullSecrets:
+    #   - myRegistryKeySecretName
+  resources: {}
+  # resources:
```
Comment (dev-958) [inline charts/kafka/values.yaml]: ditto from above

**Review by dev-958 — CHANGES_REQUESTED** [review]
Looks correct but I'd like some crud that you copy-pasta removed. Thanks.

**Review by dev-958 — DISMISSED** [review]

File: charts/kafka/charts/zookeeper/values.yaml [hunk 8]
```diff
@@ -244,6 +246,23 @@ exporters:
...
+    registry: docker.io
+    repository: bitnami/minideb
+    tag: buster
+    pullPolicy: IfNotPresent
+    ## Optionally specify an array of imagePullSecrets.
+    ## Secrets must be manually created in the namespace.
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+    ##
+    # pullSecrets:
+    #   - myRegistryKeySecretName
```
Comment (ac74475) [inline charts/kafka/charts/zookeeper/values.yaml]: remove commented out section

File: charts/kafka/charts/zookeeper/values.yaml [hunk 9]
```diff
@@ -244,6 +246,23 @@ exporters:
...
       scheme: http
 
+## volumePermissions: Change the owner of the persist volume mountpoint to RunAsUser:fsGroup
+##
+volumePermissions:
+  enabled: true
+  image:
+    registry: docker.io
+    repository: bitnami/minideb
+    tag: buster
```
Comment (ac74475) [inline charts/kafka/charts/zookeeper/values.yaml]: could we use a smaller image, this only seems to be doing a simple chown command, so maybe a busybox image at <1mb size vs this image at ~25mb

File: charts/kafka/values.yaml [hunk 10]
```diff
@@ -288,6 +291,23 @@ jmx:
...
   - kafka.log:*
 
+## volumePermissions: Change the owner of the persist volume mountpoint to RunAsUser:fsGroup
+##
+volumePermissions:
+  enabled: true
+  image:
+    registry: docker.io
+    repository: bitnami/minideb
+    tag: buster
```
Comment (ac74475) [inline charts/kafka/values.yaml]: As above I think we can reduce the image size used to <1mb by using a busybox image

File: charts/kafka/values.yaml (hunk 1 above)
Comment (ac74475) [inline charts/kafka/values.yaml:49]: Am not keen on using root here but we can tackle this later when it becomes an issue running this on OpenShift

File: charts/redis-cluster/values.yaml [hunk 11]
```diff
@@ -413,7 +413,7 @@ volumePermissions:
     registry: docker.io
     repository: bitnami/minideb
```
Comment (ac74475) [inline charts/redis-cluster/values.yaml]: Again change image to have smaller footprint

File: charts/redis/values.yaml [hunk 12]
```diff
@@ -635,7 +635,7 @@ volumePermissions:
     registry: docker.io
     repository: bitnami/minideb
     tag: buster
```
Comment (ac74475) [inline charts/redis/values.yaml]: and again

**Review by ac74475 — CHANGES_REQUESTED** [review]

File: charts/redis/values.yaml (hunk 12 above)
Comment (developer01189998819991197253) [reply charts/redis/values.yaml]: replaced

File: charts/kafka/charts/zookeeper/values.yaml [hunk 13]
```diff
@@ -244,6 +246,18 @@ exporters:
...
       scheme: http
 
+## volumePermissions: Change the owner of the persist volume mountpoint to RunAsUser:fsGroup
+##
+volumePermissions:
+  enabled: true
+  image:
+    registry: docker.io
+    repository: busybox
+    tag: latest
```
Comment (ac74475) [inline charts/kafka/charts/zookeeper/values.yaml]:
```suggestion
    tag: 1.32.0
```

File: charts/redis-cluster/values.yaml [hunk 14]
```diff
@@ -411,9 +411,10 @@ volumePermissions:
   enabled: true
   image:
     registry: docker.io
-    repository: bitnami/minideb
-    tag: buster
-    pullPolicy: Always
+    repository: busybox
+    tag: latest
```
Comment (ac74475) [inline charts/redis-cluster/values.yaml]:
```suggestion
    tag: 1.32.0
```

File: charts/redis/values.yaml [hunk 15]
```diff
@@ -633,9 +633,10 @@ volumePermissions:
   enabled: true
   image:
     registry: docker.io
-    repository: bitnami/minideb
-    tag: buster
-    pullPolicy: Always
+    repository: busybox
+    tag: latest
```
Comment (ac74475) [inline charts/redis/values.yaml]:
```suggestion
    tag: 1.32.0
```

File: charts/redis/values.yaml [hunk 16]
```diff
@@ -663,8 +664,9 @@ sysctlImage:
   enabled: false
   command: []
   registry: docker.io
-  repository: bitnami/minideb
-  tag: buster
+  repository: busybox
+  tag: latest
```
Comment (ac74475) [inline charts/redis/values.yaml]:
```suggestion
  tag: 1.32.0
```

**Review by ac74475 — CHANGES_REQUESTED** [review]

File: charts/kafka/values.yaml [hunk 17]
```diff
@@ -288,6 +291,18 @@ jmx:
...
   - kafka.log:*
 
+## volumePermissions: Change the owner of the persist volume mountpoint to RunAsUser:fsGroup
+##
+volumePermissions:
+  enabled: true
+  image:
+    registry: docker.io
+    repository: busybox
+    tag: latest
```
Comment (ac74475) [inline charts/kafka/values.yaml]:
```suggestion
    tag: 1.32.0
```

**Review by ac74475 — APPROVED** [review]

**Review by dev930018 — APPROVED** [review]

**Pull request merged**